### PR TITLE
refactor: thorchain lp positions

### PR DESCRIPTION
**Pull request by kaladinlight**
## Description

- Clean up and dry up user lp data hooks
- Add lp position type tag
- Clean up `getPoolShare`
- Fix `calculateEarnings` (now actually calculates earnings based on liquidity fees and block rewards though it is still averaged out across the entire position based on current pool share)

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

Med - refactors always bring a bit of extra risk due to amount of code moving around

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Ensure positions are still showing accurate data
    - Note: the `Earnings` value will be different.

### Engineering

:point_up: 

### Operations

:point_up:

## Screenshots (if applicable)

Earnings display (Us|ThorYield)
image

LP Position Types
image

Balance and Pool Share (Us|RuneScan)
image

### CODE REVIEW

File: src/lib/utils/thorchain/lp.ts [hunk 1]
```diff
@@ -201,7 +181,7 @@ export const estimateRemoveThorchainLiquidityPosition = async ({
     runeAmountCryptoThorPrecision,
   })
 
-  const poolShare = getPoolShare(liquidityUnitsCryptoThorPrecision, pool)
+  const poolShare = getPoolShare(pool, liquidityUnitsCryptoThorPrecision)
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/lp.ts:184]: Can we grep all those `thorPrecision` and make them `thorBaseUnit`? 🙏🏽 precision amounts refer to full-precision, not base units, so this vernacular really doesn't make sense

File: src/lib/utils/thorchain/lp.ts [hunk 2]
```diff
@@ -233,34 +213,23 @@ export const estimateRemoveThorchainLiquidityPosition = async ({
 }
 
 export const calculateEarnings = (
-  _assetLiquidityFees: string,
-  _runeLiquidityFees: string,
+  pool: MidgardEarningsHistoryPoolItem,
   userPoolShare: string,
   runePrice: string,
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/lp.ts:218]: Can we make this `runePriceUserCurrency` here to make it clear this isn't in USD?

File: src/lib/utils/thorchain/lp.ts [hunk 3]
```diff
@@ -233,34 +213,23 @@ export const estimateRemoveThorchainLiquidityPosition = async ({
...
-  assetPrice: string,
 ) => {
-  const assetLiquidityFees = fromThorBaseUnit(_assetLiquidityFees)
-  const runeLiquidityFees = fromThorBaseUnit(_runeLiquidityFees)
-
-  const userShare = bn(userPoolShare)
-  const assetEarnings = userShare.times(assetLiquidityFees).times(2).toFixed()
-  const runeEarnings = userShare.times(runeLiquidityFees).times(2).toFixed()
+  const totalEarningsRune = fromThorBaseUnit(pool.earnings).times(userPoolShare)
+  const totalEarningsFiat = totalEarningsRune.times(runePrice).toFixed()
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/lp.ts:221]: i.e not in USD
```suggestion
  const totalEarningsFiatUserCurrency = totalEarningsRune.times(runePrice).toFixed()
```

File: src/pages/ThorChainLP/Position/Position.tsx [hunk 4]
```diff
@@ -286,7 +288,7 @@ export const Position = () => {
                           <RawText>{asset?.symbol ?? ''}</RawText>
                         </Flex>
                         <Amount.Crypto
-                          value={earnings?.assetEarnings ?? '0'}
+                          value={earnings?.assetEarningsCryptoPrecision ?? '0'}
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/Position/Position.tsx:291]: 💜

File: src/pages/ThorChainLP/queries/hooks/useAllUserLpData.ts [hunk 5]
```diff
@@ -4,160 +4,101 @@ import type { UseQueryResult } from '@tanstack/react-query'
...
+            .filter(isSome)
+        },
+        select: (positions: Position[] | undefined) => {
+          return {
+            assetId,
+            positions: (positions ?? [])
+              .map(position =>
+                getUserLpDataPosition({
+                  assetId,
+                  assetPrice: marketData[assetId]?.price ?? '0',
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/queries/hooks/useAllUserLpData.ts:79]: and here too

File: src/pages/ThorChainLP/queries/hooks/useAllUserLpData.ts [hunk 6]
```diff
@@ -4,160 +4,101 @@ import type { UseQueryResult } from '@tanstack/react-query'
...
+                  assets,
+                  pool,
+                  position,
+                  runePrice: runeMarketData.price,
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/queries/hooks/useAllUserLpData.ts:83]: and there too

File: src/pages/ThorChainLP/queries/hooks/useUserLpData.ts [hunk 7]
```diff
@@ -1,17 +1,86 @@
...
+type GetPositionArgs = {
+  pool: ThornodePoolResponse
+  position: Position
+  assets: AssetsByIdPartial
+  assetId: AssetId
+  assetPrice: string
+  runePrice: string
+}
+
+export const getUserLpDataPosition = ({
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/queries/hooks/useUserLpData.ts:27]: 💜

File: src/pages/ThorChainLP/queries/hooks/useUserLpData.ts [hunk 8]
```diff
@@ -1,17 +1,86 @@
...
+  const { assetShareThorBaseUnit, runeShareThorBaseUnit, poolShareDecimalPercent } = getPoolShare(
+    pool,
+    bnOrZero(position.liquidityUnits),
+  )
+
+  const assetShareCryptoPrecision = fromThorBaseUnit(assetShareThorBaseUnit)
+  const runeShareCryptoPrecision = fromThorBaseUnit(runeShareThorBaseUnit)
+
+  const assetShareFiat = assetShareCryptoPrecision.times(assetPrice)
+  const runeShareFiat = runeShareCryptoPrecision.times(runePrice)
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/queries/hooks/useUserLpData.ts:62]: Also here

File: src/pages/ThorChainLP/queries/hooks/useUserLpData.ts [hunk 9]
```diff
@@ -124,12 +129,23 @@ export const useUserLpData = ({
...
+            assetPrice: poolAssetMarketData.price,
+            assets,
+            pool,
+            position,
+            runePrice: runeMarketData.price,
+          }),
+        )
+        .filter(isSome)
+    },
+    enabled: Boolean(assetId && currentWalletId && pool),
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/queries/hooks/useUserLpData.ts:149]: 🧠

**Review by gomesalexandre — APPROVED** [review]
Tested locally:

- Asym type tags look sane

image

- Earnings look sane against THORYield

image

- Pool share look sane against Runescan

image

- Balances look sane against develop

image

File: src/lib/utils/thorchain/lp.ts (hunk 1 above)
Comment (kaladinlight) [reply src/lib/utils/thorchain/lp.ts:184]: yessir, changes coming in next PR!

File: src/lib/utils/thorchain/lp.ts (hunk 3 above)
Comment (kaladinlight) [reply src/lib/utils/thorchain/lp.ts:221]: Similar comment here. Isn't `Fiat` generic? Just feels like `UserCurrency` is redundant imo

**Comment by kaladinlight** [comment]
closes: https://github.com/shapeshift/web/pull/6330

**Comment by kaladinlight** [comment]
Going to merge to unbork positions and can follow up re: naming based on desired outcome.

**Pull request merged**